### ISSUE TITLE
Don't use jQuery for Ajax

### ISSUE THREAD

**Issue by rod-glover** [pacificclimate/climate-explorer-frontend#45]
We're importing jQuery just to use its Ajax facility. (There are 12 references in code to the jQuery alias `$`, and they're all `$.ajax()` calls.)

Using jQuery for this purpose seems a little perverse:

1. jQuery is not mainly about dispatching HTTP requests. That's a lot of overhead to get an HTTP package.
2. The promise-like objects jQuery returns [are not standards-compliant promises](https://blog.domenic.me/youre-missing-the-point-of-promises/), and the differences are [confusing and troublesome](http://stackoverflow.com/questions/23744612/problems-inherent-to-jquery-deferred-jquery-1-x-2-x). And we're using these promisey objects, and I think [already hitting trouble](https://github.com/pacificclimate/climate-explorer-frontend/pull/44#discussion_r117149437).
3. [There are HTTP libraries that return standards-compliant promises](https://daveceddia.com/ajax-requests-in-react/), `axios` and `fetch` (itself likely to become a standard) for two.

If we want to use promises for async operations (and I think it's a good idea, though [not for all applications](http://sealedabstract.com/code/broken-promises/)), may I suggest we retool away from jQuery and its icky promisey things before we have much more code written around it? A quick perusal of the CE frontend code suggests this won't be hard, and it could save us some headaches later.

**Comment by jameshiebert** [comment]
I'd be happy to entertain a transition to a better library for AJAX requests, especially if you have experience with a couple of them. Do you want to take this on and then share your knowledge with @corviday (to be incorporated into #44)?

**Comment by rod-glover** [comment]
Happily. I have experience with the HTTP request component provided by Angular, which returns Q (standards-compliant) promises, and with another promise-compliant library whose identity I don't know because it was just a given in the large codebase I was working on at Change.org.

I'll do a little research on the currently available options.

**Comment by rod-glover** [comment]
Research

A basic question arose: Do we want to adopt an HTTP library with a callback style API and wrap it with promises, or do we want to adopt a library that returns promises directly? Ideally we'd have both, but since the only library to offer both (`reqwest`) looks immature and under-maintained, we have to choose between the two options.

In response to a [popular blog post by James Coglan on the virtues of promises](https://blog.jcoglan.com/2013/03/30/callbacks-are-imperative-promises-are-functional-nodes-biggest-missed-opportunity/), a [post by Drew Crawford outlines the shortcomings of promises](http://sealedabstract.com/code/broken-promises/), which is respectfully considered in [a follow up article by Coglan](https://blog.jcoglan.com/2013/04/01/callbacks-promises-and-simplicity/). Boiled down, Crawford's points are: 

- a promise never releases the resource it represents, and in some cases that can cause excessive memory usage
- some asynchronous applications call for much greater control over control flow than promises permit (the point of promises is that you delegate control flow to the promise library)

Assumption: It appears that CE does not load many large objects, nor does it have tricky asynchronous situations. Therefore these objections do not appear to be blockers for adopting a promises-only library here.

Given that assumption, a promises-based library is preferable in order to reduce overhead and the opportunities for introducing errors.

With that question settled, we can proceed to criteria for selecting a library for this application:

1. (essential) works in Chrome & Firefox
2. (essential?) works in "all" browsers
3. (essential) installable in Node.js so that it can be webpacked
4. (essential) supports standards compliant promises
5. (desirable) option to use callbacks for cases where Promises are not appropriate
6. (desirable) test support for easy mocking/replaying etc.

The following table (pasted as an image) summarizes the results. This table extends the useful info in [this blog post on Ajax libraries](http://andrewhfarmer.com/ajax-libraries/).

![image](https://cloud.githubusercontent.com/assets/4139706/26259485/18896f6e-3c7e-11e7-84fd-746eefa77510.png)

**Comment by rod-glover** [comment]
# Recommendation

TL;DR: Adopt `axios`.

Two candidates meet the essential criteria: 

- `whatwg-fetch` (a polyfill for the `fetch` API under consideration as a new ES6+ standard)
- `axios`

Of the two, `axios` wins on test/debugging support, and it also offers a fuller API than `whatwg-fetch`.
